### PR TITLE
[TESTS] Timezone awareness for history datetime comparison

### LINKED ISSUE #90: [TESTS] test_browsers_* outcome is timezone-dependent
Hi, whilst working to complete a PR (for Issue #65) I noticed that the `test_browsers.py` tests were all failing, though all tests passed for `test_generic` and `test_imports`. The errors reported indicate a datetime mismatch and after some investigation, it seems the failures only occur when the tests are run in specific timezones (see the detail, including outputs, provided below). Overall, this seems to imply that either:
1. the datetimes hard-coded into that test fixture are not processed properly to account for any possible timezone of a machine running the tests;
2. there is a timezone-related bug in the datetimes returned in the browser history tuples.

I have not yet determined which of these is the case, if any. I am happy to look into it further, though, assuming it is not obvious to you what the issue is here, and ultimately put in a PR to fix either the tests or the bug (as appropriate) towards hacktoberfest if that is okay? In which case, it would be useful to know if others can recreate these failures by temporarily re-setting the timezone of their machine and running the tests.

# Detail

Note: I am running `browser-history` after installing it in development mode i.e. from the current `master` (via `pip install -e .`).

If I run `pytest` from the root repo directory `browser-history`, I observe the following:

```pytb
============================= test session starts ==============================
platform linux -- Python 3.7.6, pytest-5.3.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/sadie/browser-history
plugins: arraydiff-0.3, doctestplus-0.5.0, openfiles-0.4.0, astropy-header-0.1.2, cov-2.10.1, remotedata-0.3.2, hypothesis-5.24.0
collected 13 items                                                             

tests/test_browsers.py FFFFFFF                                           [ 53%]
tests/test_generic.py .....                                              [ 92%]
tests/test_import.py .                                                   [100%]

=================================== FAILURES ===================================
______________________________ test_firefox_linux ______________________________

become_linux = 'Linux', change_homedir = 'Linux'

    def test_firefox_linux(become_linux, change_homedir):  # noqa: F811
        """Test history is correct on Firefox for Linux"""
        f = browser_history.browsers.Firefox()
        output = f.fetch_history()
        his = output.histories
        assert len(his) == 5
>       assert his[0] == (
            datetime.datetime(
                2020,
                8,
                3,
                0,
                29,
                4,
                tzinfo=datetime.timezone(datetime.timedelta(seconds=19800), "IST"),
            ),
            "https://www.mozilla.org/en-US/privacy/firefox/",
        )
E       AssertionError: assert (datetime.dat...acy/firefox/') == (datetime.dat...acy/firefox/')
E         At index 0 diff: datetime.datetime(2020, 8, 2, 19, 59, 4, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')) != datetime.datetime(2020, 8, 3, 0, 29, 4, tzinfo=datetime.timezone(datetime.timedelta(seconds=19800), 'IST'))
E         Use -v to get the full diff

tests/test_browsers.py:21: AssertionError
______________________________ test_chrome_linux _______________________________

become_linux = 'Linux', change_homedir = 'Linux'

    def test_chrome_linux(become_linux, change_homedir):
        """Test history is correct on Chrome for Linux"""
        f = browser_history.browsers.Chrome()
        output = f.fetch_history()
        his = output.histories
        assert len(his) == 2
>       assert his[0] == (
            datetime.datetime(
                2020,
                10,
                15,
                15,
                34,
                30,
                tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), "CEST"),
            ),
            "www.github.com",
        )
E       AssertionError: assert (datetime.dat...w.github.com') == (datetime.dat...w.github.com')
E         At index 0 diff: datetime.datetime(2020, 10, 15, 14, 34, 30, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')) != datetime.datetime(2020, 10, 15, 15, 34, 30, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), 'CEST'))
E         Use -v to get the full diff

tests/test_browsers.py:57: AssertionError
_____________________________ test_chromium_linux ______________________________

become_linux = 'Linux', change_homedir = 'Linux'

    def test_chromium_linux(become_linux, change_homedir):
        """Test history is correct on Chromium for Linux"""
        f = browser_history.browsers.Chromium()
        output = f.fetch_history()
        his = output.histories
        assert len(his) == 2
>       assert his[0] == (
            datetime.datetime(
                2020,
                10,
                15,
                15,
                34,
                30,
                tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), "CEST"),
            ),
            "www.github.com",
        )
E       AssertionError: assert (datetime.dat...w.github.com') == (datetime.dat...w.github.com')
E         At index 0 diff: datetime.datetime(2020, 10, 15, 14, 34, 30, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')) != datetime.datetime(2020, 10, 15, 15, 34, 30, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), 'CEST'))
E         Use -v to get the full diff

tests/test_browsers.py:94: AssertionError
_____________________________ test_firefox_windows _____________________________

become_windows = 'Windows', change_homedir = 'Windows'

    def test_firefox_windows(become_windows, change_homedir):  # noqa: F811
        """Test history is correct on Firefox for Windows"""
        f = browser_history.browsers.Firefox()
        output = f.fetch_history()
        his = output.histories
        assert len(his) == 8
        profs = f.profiles(f.history_file)
        assert len(profs) == 2
        # the history list is long so just check the first and last item
>       assert his[0] == (
            datetime.datetime(
                2020,
                10,
                1,
                11,
                43,
                35,
                tzinfo=datetime.timezone(
                    datetime.timedelta(seconds=10800), "E. Africa Standard Time"
                ),
            ),
            "https://www.youtube.com/",
        )
E       AssertionError: assert (datetime.dat...youtube.com/') == (datetime.dat...youtube.com/')
E         At index 0 diff: datetime.datetime(2020, 10, 1, 9, 43, 35, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')) != datetime.datetime(2020, 10, 1, 11, 43, 35, tzinfo=datetime.timezone(datetime.timedelta(seconds=10800), 'E. Africa Standard Time'))
E         Use -v to get the full diff

tests/test_browsers.py:137: AssertionError
______________________________ test_edge_windows _______________________________

become_windows = 'Windows', change_homedir = 'Windows'

    def test_edge_windows(become_windows, change_homedir):  # noqa: F811
        """Test history is correct for Edge on Windows"""
        e = browser_history.browsers.Edge()
        output = e.fetch_history()
        his = output.histories
        # test history from all profiles
        assert len(his) == 1
>       assert his == [
            (
                datetime.datetime(
                    2020,
                    9,
                    23,
                    10,
                    45,
                    3,
                    tzinfo=datetime.timezone(
                        datetime.timedelta(seconds=19800), "India Standard Time"
                    ),
                ),
                "https://pesos.github.io/",
            ),
        ]
E       AssertionError: assert [(datetime.da....github.io/')] == [(datetime.da....github.io/')]
E         At index 0 diff: (datetime.datetime(2020, 9, 23, 6, 15, 3, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')), 'https://pesos.github.io/') != (datetime.datetime(2020, 9, 23, 10, 45, 3, tzinfo=datetime.timezone(datetime.timedelta(seconds=19800), 'India Standard Time')), 'https://pesos.github.io/')
E         Use -v to get the full diff

tests/test_browsers.py:193: AssertionError
_______________________________ test_safari_mac ________________________________

become_mac = 'Darwin', change_homedir = 'Darwin'

    def test_safari_mac(become_mac, change_homedir):  # noqa: F811
        """Test history is correct for Safari on macOS"""
    
        e = browser_history.browsers.Safari()
        output = e.fetch_history()
        his = output.histories
        assert len(his) == 5
>       assert his[0] == (
            datetime.datetime(
                2020,
                9,
                29,
                23,
                34,
                28,
                tzinfo=datetime.timezone(datetime.timedelta(seconds=19800), "IST"),
            ),
            "https://www.apple.com/in/",
        )
E       AssertionError: assert (datetime.dat...pple.com/in/') == (datetime.dat...pple.com/in/')
E         At index 0 diff: datetime.datetime(2020, 9, 29, 19, 4, 28, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')) != datetime.datetime(2020, 9, 29, 23, 34, 28, tzinfo=datetime.timezone(datetime.timedelta(seconds=19800), 'IST'))
E         Use -v to get the full diff

tests/test_browsers.py:245: AssertionError
______________________________ test_opera_windows ______________________________

become_windows = 'Windows', change_homedir = 'Windows'

    def test_opera_windows(become_windows, change_homedir):
        o = browser_history.browsers.Opera()
        outputs = o.fetch_history()
        his = outputs.histories
        assert len(his) == 2
>       assert his == [
            (
                datetime.datetime(
                    2020,
                    10,
                    13,
                    12,
                    4,
                    51,
                    tzinfo=datetime.timezone(
                        datetime.timedelta(seconds=10800), "E. Africa Standard Time"
                    ),
                ),
                "https://www.youtube.com/",
            ),
            (
                datetime.datetime(
                    2020,
                    10,
                    13,
                    12,
                    4,
                    59,
                    tzinfo=datetime.timezone(
                        datetime.timedelta(seconds=10800), "E. Africa Standard Time"
                    ),
                ),
                "https://github.com/",
            ),
        ]
E       AssertionError: assert [(datetime.da...github.com/')] == [(datetime.da...github.com/')]
E         At index 0 diff: (datetime.datetime(2020, 10, 13, 10, 4, 51, tzinfo=datetime.timezone(datetime.timedelta(0), 'GMT')), 'https://www.youtube.com/') != (datetime.datetime(2020, 10, 13, 12, 4, 51, tzinfo=datetime.timezone(datetime.timedelta(seconds=10800), 'E. Africa Standard Time')), 'https://www.youtube.com/')
E         Use -v to get the full diff

tests/test_browsers.py:277: AssertionError
========================= 7 failed, 6 passed in 0.17s ==========================
```

But I notice that if I change the timezone I have set on my machine via:

```console
$ sudo timedatectl set-timezone America/New_York
```

I then see the tests all pass:

```pytb
============================= test session starts ==============================
platform linux -- Python 3.7.6, pytest-5.3.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/sadie/browser-history
plugins: arraydiff-0.3, doctestplus-0.5.0, openfiles-0.4.0, astropy-header-0.1.2, cov-2.10.1, remotedata-0.3.2, hypothesis-5.24.0
collected 13 items                                                             

tests/test_browsers.py .......                                           [ 53%]
tests/test_generic.py .....                                              [ 92%]
tests/test_import.py .                                                   [100%]

============================== 13 passed in 0.19s ==============================
```

I thought I would try some other timezones out to see which might lead to failures. I found, by changing machine timezone via the equivalent command to the above, that all the `test_browser` tests passed for timezones such as:

* `UTC`
* `Asia/*` for several cities I tried e.g. `Tokyo`, `Calcutta`;

but the following led to varying numbers of failures:

* `Europe`/*` for cities I tried including `London` and `Sofia` (all 7 fail);
* `Pacific/*` e.g . as tested with `Auckland` (2/7 fail);
* `Australia/*` e.g. `Melbourne` (4/7 fail).

Usually but not always the failure `diff` indicated a mismatch between the expected and actual datetimes of one hour so it could be something to do with daylight savings. I am not sure so far. I do however notice that there was a commit message indicating that timezones had been a problem at least on GH Actions (41fee0654d08605513e895f41cb585e9665cd29c) at some point, but that specifically refers to UTC which as I saw was a timezone specification where all the tests passed, and the bug if any, was not present.

### PR DESCRIPTION
**Pull request by sadielbartholomew**
# Description

Close #90 by making the `test_browser` history comparisons check for equivalence of datetimes in a way that is properly timezone aware due to use of the `dateutil.tz` module.

I wanted to avoid the need for any external module even in the test modules, given that the core codebase of `browser-history` has zero dependencies, but unfortunately after much investigation I found it was not possible to solve the issue without importing a truly timezone-aware library (`pytz` is another that will work) into the test in question. However, in Python 3.9 there is a built-in solution `zoneinfo` which can be used, so in future the new dependency can be removed with a tweak of the test to use that instead.

However, given the tests themselves require the `pytest` external library to run, I think it should be okay to add one dependency to the test requirements? I should stress again that it really isn't possible to solve the issue at the moment without an external library that manages timezones.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have read the [contribution guidelines](https://browser-history.readthedocs.io/en/latest/contributing.html) and followed it as far as possible. 
- [x] I have performed a self-review of my own code (if applicable)
- [x] I have commented my code, particularly in hard-to-understand areas (if applicable)
- [n/a] I have made corresponding changes to the documentation (if applicable)
- [x] My changes generate no new warnings 
- [x] I have enabled the pre-commit hook and it's not detecting any issue.
- [x] Any dependent and pending changes have been merged and published

### CODE REVIEW

**Comment by sadielbartholomew** [comment]
The CI tests are failing because they do not have the `dateutil` module installed on the Actions job VMs. I'll add the installation of it to the GH Actions workflow assuming you are happy to add it as a test suite (only) dependency, but I will need confirmation it is okay.

I was also wondering whether it would be better to add another commit to consolidate the comparisons further to prevent some duplication of the timezone-processing logic, going from the general form:

```python
LOCAL_TIMEZONE = tz.gettz()

...

assert _detatch_timezone_stamp(his[0]) == (
    datetime.datetime(
        <datetime tupe ints>,
        tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), "CEST"),
    ),
    )
    .astimezone(LOCAL_TIMEZONE)
    .replace(tzinfo=None),
    <url>,
)
```
to instead delegate to a function to handle the assertions. Something like:

```python
def assert_histories_equal_in_any_timezone(expected_history, actual_history):
    """TODO."""
    local_timezone = tz.gettz()

    assert (
        _detatch_timezone_stamp(expected_history[0]) ==
        (actual_history[0].astimezone(local_timezone).replace(tzinfo=None),
         actual_history[1])
    )
```

however it might be less explicit i.e. not obvious the method holds an assertion in each test method. If you prefer that let me know, else I am happy to leave it in the current form.

**Review by Samyak2 — COMMENTED** [review]
Hi. Thank you for the PR!

> However, given the tests themselves require the `pytest` external library to run, I think it should be okay to add one dependency to the test requirements? I should stress again that it really isn't possible to solve the issue at the moment without an external library that manages timezones.

Yes, as long as the dependency is only for test suite it is ok.

> The CI tests are failing because they do not have the `dateutil` module installed on the Actions job VMs. I'll add the installation of it to the GH Actions workflow assuming you are happy to add it as a test suite (only) dependency, but I will need confirmation it is okay.

Yes, you can go ahead and add it to the workflow.

> to instead delegate to a function to handle the assertions. Something like:

That looks good, it would make the comparison more modular. These functions should go into `tests/utils.py` I think and maybe rename `assert_histories_equal_in_any_timezone` to just `assert_histories_equal`.

File: tests/test_browsers.py [hunk 1]
```diff
@@ -9,6 +9,38 @@
...
+
+
+# Instead setting this to the standard library timezone option:
+#     datetime.datetime.now(datetime.timezone.utc).astimezone().tzinfo
+# will cause failures for timezones e.g. 'Europe/London'
+# or 'Australia/Melbourne' due to DST effects.
+LOCAL_TIMEZONE = tz.gettz()
+
+
+def _detatch_timezone_stamp(hist):
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply tests/test_browsers.py:25]: Well spotted, thanks. I'll get that fixed.

**Comment by sadielbartholomew** [comment]
Hi @Samyak2, thanks for the comments. I'm very busy until tomorrow afternoon (my time, UK) but by Wednesday I will update this to address your feedback.

**Comment by sadielbartholomew** [comment]
Sorry for being later than suggested in my above comment, @Samyak2. I have now updated this PR to address all of your feedback, so as far as I am concerned this is ready to merge. (That said, I am happy to make further amendments to the PR if you wish.)

**Review by Samyak2 — APPROVED** [review]
Looks good.
Thank you!
